purple-pill/js/controllers/fontexportController.js

```diff
@@ -20,152 +92,136 @@ function($scope, $http, sharedScope, $timeout) {
+        var percentage
+          , currentGlyph = it.value['current_glyph']
+          , totalGlyphs = it.value['total_glyphs']
+          ;
+        percentage = 100.0 * (instanceIndex + (currentGlyph/totalGlyphs)) / totalInstances;
+        return percentage;
+    }
 
-        function get_timestamp(){
-            var year, month, day, hours, minutes, seconds
-              , date = new Date()
-              ;
-            year = zero_padding(date.getFullYear());
-            month = zero_padding(date.getMonth() + 1); //getMonth() returns a zero-based value: 0=January, 11=December
-            day = zero_padding(date.getDate());
-            hours = zero_padding(date.getHours());
-            minutes = zero_padding(date.getMinutes());
-            seconds = zero_padding(date.getSeconds());
-
-            return [year, month, day].join("") + "-" + [hours, minutes, seconds].join("");
-        }
-
+    function exportFont_compute_glyphs(exportObjects) {
         var bundle = new $scope.data.stateless.JSZip()
-          , bundleFolderName = "metapolator-export-" + get_timestamp()
-          , bundle_filename = bundleFolderName + ".zip"
+          , bundleFolderName = "metapolator-export-" + getTimestamp()
+          , bundleFileName = bundleFolderName + ".zip"
           , bundleFolder = bundle.folder(bundleFolderName)
-          , bundle_data
-          , instances_for_export = Array()
+          , totalInstances = exportObjects.length
+          , progress = new ProgressBar( $("#progressbar")
+                                      , $("#progresslabel")
+                                      , UI_UPDATE_TIMESLICE )
           ;
-        angular.forEach($scope.data.families, function(family) {
-            angular.forEach(family.instances, function(instance) {
-                if (instance.exportFont){
-                    instances_for_export.push(instance);
-                }
-            });
-        });
+        function _exportFontComputeGlyphs(exportObjects, totalInstances, bundleFolder, resolve, reject){
+            if (exportObjects.length==0){
+                resolve(true);
+                return;
+            }
 
-        function getExportGenerator(index){
-            var instance = instances_for_export[index]
-              , exportObjects = $scope.data.stateful.project.getUFOExportGenerator(
-                    instance.name,
-                    instance.displayName + ".ufo",
-                    /* precision: */ -1
-                )
-              , gen = exportObjects[0]
+            var text
+              , percentage
+              , index = totalInstances - exportObjects.length
+              , obj = exportObjects[exportObjects.length - 1]
+              , it = obj.generator.next()
+              , data, name
               ;
-            io[index] = exportObjects[1];
-            return gen;
-        }
+            if (!it.done){
+                text = exportingGlyphMessage(it, index, totalInstances);
```

Comment:
```js
if (!it.done) {
    // this makes it easy for us to use another progress api, or even none if we choose so
    // also we don't need to know anything about updating the progress, it can send short messages if it choses
    if(progress)
        progress.setData(index, totalInstances, it.value);
}
```